# Lesson 0: Working on TIKE, with Cloud Data

## Learning Goals: 
- Learn about TIKE and working on a cloud platform.
- Define cloud terminology: what is a “bucket” or a server? For that matter, what is the “cloud”?
- Access data through astroquery by name, region, or criteria
- Download TESS data and show an image

<!-- ## Lesson Outline:
- Go over TIKE: what can it do? What are its limits (memory, processing power, etc.)?
- Cloud overview. What does it mean to work “on the cloud”? How is this different from working on a laptop? [Provide motivation: why are we doing this?]
- Demo cloud data access using astroquery. Meaningfully interact with data (maybe just read FITS headers) so people see they don’t need to perform a download.
 -->

## Prerequisites
If you are unfamiliar with Python, now is your chance to make sure you understand the basics before continuing. Basics include: 
- installing and importing packages (conda or pip)
- data object types (floats, ints, strings, etc.)
- arrays and lists
- for loops and if statements
#### We will use these things but may not explain them in depth.

## What is TIKE?

TIKE stands for the Time series Integrated Knowledge Engine.

TIKE uses a web-based platform, called JupyterHub, which enables users to run Python Notebooks and other software remotely in a web browser without needing to install anything on your local computer. Notebooks are a convenient way of packaging code, its outputs, and visualizations. TIKE is intended for astronomical researchers (or students like you!) to access and analyze data from NASA's Kepler or TESS missions. You can find more info on TESS [here](https://archive.stsci.edu/missions-and-data/tess#section-33e8b9e7-9cc2-414a-b830-c88f2b409b13). 

TIKE is continually maintained and updated by real humans, so if you run into issues please don't hesistate to let us know or send us your suggestions for packages and tutorials. You can do that either through the [MAST help desk](mailto:archive@stsci.edu) or [opening an issue on GitHub](https://github.com/spacetelescope/tike_content).

<!-- #### Wait...what is Github? 
GitHub is a code hosting platform for version control and collaboration. It lets you and others work together on coding projects from anywhere. And you're in luck! Github hosts a beginnner course on how to get started here: https://github.com/skills/introduction-to-github.  -->

## What is the "cloud"?

The "cloud" is a term used to describe a global network of servers, each with a unique function. The cloud is not a physical entity, but instead is a vast network of remote servers around the globe which are hooked together and meant to operate as a single data ecosystem. In other words, the cloud refers to servers that are accessed over the internet, and the software and databases that run on those servers. 

#### So...whats a server? 
A server is a computer or computer program which manages access to a centralized resource or service in a network. This can be a physical computer somewhere, or a virtual instance. Think of the server as the waiter in a restaurant, where your data is the food. Like in a restaurant, if there are a lot of people trying to order food (or query data) at one time, it can overwhelm the server and the service will be slower. 

Additionally, inside a cloud you'll also find something called a "bucket", these are basically the data structures that hold your data on the cloud, all the data on a cloud must be contained in a bucket. You can use buckets to organize your data and control access to your data, but unlike directories and folders, you cannot nest buckets.

#### Why would I want to work on the cloud?
Using the cloud has several benefits, the major one being that you can access data whenever and wherever you want to, from any device, as long as you have WiFi. Additionally, working on the cloud means you don't have to download as much data to your machine, saving memory space and run time. 


Some common examples of clouds that one can host data on are: 
   - Amazon Web Services (AWS)
   - Google Cloud
   - Microsoft Azure
   - IBM Cloud

The Mikulski Archive for Space Telescopes (MAST) data archive provides a copy of several large NASA data sets via Amazon Web Services (AWS) as part of the MAST AWS Public Datasets program. The data provided by MAST is free to use and download, no need for an AWS account!

### What's the difference between working on the cloud and working on TIKE?
TIKE allows you to acess data without having to download it to your local machine, everything you do and use can be stored on the TIKE server, it is the platform through which you can access cloud data. In essense, you don't need to understand how cloud computing works in order to utulize its power on TIKE. TIKE makes things as easy as opening up a Jupyter Notebook! 

## How do I access the cloud?
There are many options for accessing data on the cloud, including command line tools and various Python packages. For this lesson, we will be using a Python data query package called Astroquery. 

Astroquery is a Python package inside the Astropy Project. The Astropy Project is a community effort to develop a common core package for astronomy in Python and fosters an ecosystem of interoperable astronomy packages. Astroquery is a package that allows users to access a variety of data archives for the purposes of astronomical research. We will use it to access TESS data on the cloud hosted by MAST. 


## Astroquery demonstration

### Imports
The following cell holds the imported packages. These packages are necessary for running the rest of the cells in this notebook, and you can expect to use some of these packages almost everytime you do astronomical research. A description of each import is as follows:

* `numpy` to handle array mathematics
* `pandas` to handle date conversions
* `fits` from astropy.io for accessing FITS files
* `Table` from astropy.table for creating tidy tables of the data
* `WCS` from astropy.wcs for storing World Coordinate Systems information 
* `SkyCoord` from astropy.coordinates for defining RA and Dec for targets
* `matplotlib.pyplot` for plotting data and images
* `Mast` and `Observations` from astroquery.mast for querying data and observations from the MAST archive

In [6]:
from astropy.io import fits
from astropy.table import Table
from astropy.wcs import WCS
from astropy.coordinates import SkyCoord
from astroquery.mast import Mast, Observations
from astropy.io import ascii
import requests
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

### Querying TESS data from MAST
We will be using the Observations class in the astroquery.mast subpackage from Astroquery. We will show you a few different ways to access data on an object in the TESS data products. 
Accessing data can be done with the following functions:
- query_object( )
- query_region( )
- query_criteria( )

All of which provide different levels of scrunity for the data you would like to retrieve. 


### Choosing a target
First, we have to choose star that you'd like to get TESS data of. We can query the MAST cloud server using just the name of the target with the query_object( ) function. It is best practice to reduce the size of the data you are accessing by selecting only data within a certain radius of the target. 

Let's start with the star Fomalhaut, the brightest star in the southern constellation of Piscis Austrinus, the "Southern Fish", and one of the brightest stars in the night sky.

<img src="https://upload.wikimedia.org/wikipedia/commons/a/ae/Heic0821f.jpg" width="300">

It is located at Right ascension: 22h 57m 39.0465s, Declination: −29° 37′ 20.050″ with an aparant magnitude in the V-band of 1.16. 


### query_object( )
We can query the MAST cloud server using just the name of the target with the query_object( ) function. It is best practice to reduce the size of the data you are accessing by selecting only data within a certain radius of the target. 


In [ ]:
#Query MAST for TESS observations of Fomalhaut 
obs = Observations.query_object("Fomalhaut", radius="1s")
# want = (obs['obs_collection'] == "TESS") & (obs['sequence_number'] == 29) & (obs['dataproduct_type'] == 'timeseries')

#Print out the observations we have queried 
obs['obs_collection', 'project','obs_id']

#Let's parse these observations for just those from TESS
want = (obs['obs_collection'] == "TESS")

#Print out the observations we have queried 
obs[want]['obs_collection', 'project','obs_id']


<div class="alert-success">
    <b>Using this query method, how many TESS observations are within 1 arcsecond of Trappist-1?</b>
</div>

ANS: 
<div class="alert-danger">
    <b></b>
</div>

### query_region( )
This function allows us to choose a specific region of the sky and query all the data that TESS has inside that region, defined by some radius from a set of coordinates. So, we will use the coordinates for Fomalhaut with a 2 arcsecond radius around it. 

TESS requires your coordinates to be converted into degrees. For the Right Ascension, we know that there are 15° in each hour (24 hr a day to rotate 360° = 15° per hour), so let's convert everything to hours and then mulitply by 15. We can do this by dividing our seconds by 3600, dividing our minutes by 60, leaving the hours alone, and adding them all up. For the declination, we can convert to degrees by dividing our arcseconds by 3600, dividing our arcminutes by 60, leaving the degrees alone, and then adding all three values together. So, for Fomalhaut this would be:

$ RA = (22h + 57m/60 + 39.0465s/3600) * 15 = 344.41269272° $

$ DEC = -29° + 37'/60 + 20.050"/3600 = -29.62223703° $

#### Packages for unit conversion
With all that being said, there are robust packages out there for unit conversions. Astropy.units is capable of storing variables with units built-in to the data object and will provide simple conversion options. We will provide an example of this down the line.

In [3]:
coordinates = SkyCoord(344.41269272, -29.62223703, unit="deg") #defaults to ICRS frame

observations = Observations.query_region(coordinates, radius = "2s")
#Parse the query for only TESS observations
obs_wanted = (observations['obs_collection'] == 'TESS') #observations['dataproduct_type'] == 'timeseries')

#Let's take a peek at the data
print( observations[obs_wanted]['obs_collection', 'project','obs_id'] )

obs_collection project                      obs_id                    
-------------- ------- -----------------------------------------------
          TESS    TESS                                  tess-s0002-1-1
          TESS    TESS                                  tess-s0029-1-1
          TESS    TESS tess2018234235059-s0002-0000000047552789-0121-s
          TESS    TESS tess2020238165205-s0029-0000000047552789-0193-s


 <div class="alert-success">
    <b>Try to query a larger region, does TESS have more observations when the search radius is larger?</b>
</div>

ANS: 
<div class="alert-danger">
    No, TESS only has 4 observations on this region of the sky, even when the search radius is made larger around these coordinates, no other observations show up. 
</div>

 <div class="alert-success">
    <b>In what scenarios would query_region be a better choice than query_object?</b>
</div>

ANS: 
 <div class="alert-danger">
    query_region allows you to not specify the object name if you only have or would only like to input the coordinates with a search radius. query_object is just a wrapper function for this which allows you to use an object name if you have one. 
</div>

### query_criteria( )
This function allows us to give astroquery a larger number of parameters to refine the data to exactly what we want in just one line of code. 

In [4]:
TESS_table = Observations.query_criteria(objectname="Fomalhaut",radius="2 arcsec", obs_collection="TESS") #dataproduct_type="timeseries"

# Let's print out some relevant columns of this table
columns = ["instrument_name","filters","target_name","obs_id","calib_level","t_exptime"]
TESS_table[columns].show_in_notebook(display_length=10)

<div class="alert-success">
    <b>How could you tell if the query produced no data?</b>
</div>

ANS: 
<div class="alert-danger">
    The table length would = 0
</div>

<div class="alert-success">
    <b>Which query function is the most general/broad/unspecific/etc. (to choose right word)? How can you tell?</b>
</div>

ANS: 
<div class="alert-danger">
    query_object is the most general, if you don't 
    parse the results it will return any observation that MAST has from any mission for that target name (not just TESS). 
</div>

### Plot a Full Frame Image
We can see from our observation table that some of the data TESS has on Fomalhaut are Full Frame Images. These will be stored in FITS files which can be downloaded and viewed on TIKE. 

#### What is a FITS file?
File Image Transport System (FITS) is a file format designed to store, transmit, and manipulate scientific images and associated data. It is the most widely used file type in astronomical research. A FITS file consists of one or more Header + Data Units (HDUs), where the first HDU is called the primary HDU, or primary array. The primary array may be empty or contain an N-dimensional array of pixels, such as a 1-D spectrum, a 2-D image, or a 3-D data cube. 

For more info on FITS files, here is a link to the FITS official [website](https://fits.gsfc.nasa.gov/). 

In [5]:
# Query the observations from MAST to get a list of products for our selected observations
data_products = Observations.get_product_list(TESS_table)

# Get the minimum required products
filtered = Observations.filter_products(data_products, dataproduct_type ='image')

#check out what this filter returned
filtered

<Table masked=True length=9760>
 obsID   obs_collection dataproduct_type ... dataRights calib_level
  str8        str4           str10       ...    str6       int64   
-------- -------------- ---------------- ... ---------- -----------
60896979           TESS            image ...     PUBLIC           2
60896979           TESS            image ...     PUBLIC           1
60887674           TESS            image ...     PUBLIC           2
60887674           TESS            image ...     PUBLIC           1
60895527           TESS            image ...     PUBLIC           2
60895527           TESS            image ...     PUBLIC           1
     ...            ...              ... ...        ...         ...
27958143           TESS            image ...     PUBLIC           1
27919138           TESS            image ...     PUBLIC           2
27919138           TESS            image ...     PUBLIC           1
27939489           TESS            image ...     PUBLIC           2
27939489           TESS            image ...     PUBLIC           1
27953469           TESS            image ...     PUBLIC           2
27953469           TESS            image ...     PUBLIC           1

Let's plot one of the calibrated full frame images of Fomalhaut from TESS. We can choose a Calibrated FFI data product at random, how about obsID 60895527. Let's see how that one looks!

We need to filter our data again so that we don't have to download all 9760 data products, we just want the data for obsID 60895527 with a description of "Calibrated full frame image". 

In [ ]:
filtered2 = Observations.filter_products(data_products, dataproduct_type ='image', obsID = '60895527', description = 'Calibrated full frame image')

#Print out the data products we have filtered for
filtered2

Download the FITs file

In [ ]:
data = Observations.download_products(filtered2)
#Take a peek at the FITS file we downloaded
fits_file = data['Local Path'][0]

### Understanding the FITS FFI structure

TESS FFI FITS files contain a primary HDU with metadata stored in the header.  The first extension HDU contains more metadata in the header, and stores the full frame image.  The second extension HDU contains the uncertainty values for the image.  Let's examine the structure of the FITS file using the astropy.fits `info` function, which shows the FITS file format in more detail.

In [ ]:
fits.info(fits_file)

<div class="alert-success">
    <b>What are the dimensions of the Full Frame Image?</b>
</div>

ANS:
<div class="alert-danger">
    <b>2136 pixels by 2078 pixels</b>
</div>

### Reading the WCS and Calibrated Image

Now that we have the file, let's store the world coordinate system information for use later. World Coordinate Systems (WCSs) describe the geometric transformations between one set of coordinates and another, almost all the FITS files you come across will have a WCS in its metadata. We can use the astropy.wcs WCS function to store the information from the image extension HDU's header. 

In [ ]:
# open the file, extract the WCS and Image data, and then close the file
with fits.open(fits_file, mode = "readonly") as hdulist:
    wcs = WCS(hdulist[1].header)
    image = hdulist[1].data
    header = hdulist[1].header
    
#Take a peek at the header
header

### Display the Image

Finally, let's plot our full frame image of Fomalhaut. 

In [ ]:
plt.figure(figsize = (12,12))
# plt.subplot(projection = wcs) #TO FIX -this is not working, i think its something to do with my versions of matplotlib and astropy
plt.imshow(image, vmin = np.percentile(image,4),vmax = np.percentile(image, 98),origin = "lower", cmap = 'inferno')
plt.ylim(1100,1900)
plt.xlim(200,1000)
plt.xlabel('RA')
plt.ylabel('Dec')
plt.show()

## Homework
1. Choose a star from this list:
    - Proxima Centauri
    - Procyon (* alf CMi)
    - Aldebaran (* alf Tau)
    - Polaris (alf UMi)
    - Rigel (* bet Ori)
    
Next, query TESS observations for it, download a fits file for 1 calibrated FFI, and then print out the FITS Header. Answer the following questions using the code above as an example. HINT: you may need to find the coordinates for your chosen star, [SIMBAD](https://simbad.u-strasbg.fr/simbad/sim-fid) is a reliable source. 


- What is the date of this observation?
- What are CRVAL1 and CRVAL2?
- How many TESS observations of this star are there?
- What are the exposure times of these observations?

2. Using the data you queried, plot a FITS FFI for the star. 

## Additional Resources
Can't get enough? Here are some links to more information!

If you need an introduction (or a refresher!) to basic Python syntax, there are several great resources available online. [CodeAcademy](https://www.codecademy.com/learn/learn-python-3) is a great service with a totally free option for getting started with Python, note you will have to create an account to use it. Additionally, the Youtube channel FreeCodeCamp.org has a great [video tutorial](https://www.youtube.com/watch?v=rfscVS0vtbw) on everyting you need to get started programming in Python. Another good resource is the [Python 4 Everyone](https://www.py4e.com/) book. 

The full astropy documentation can be found [here](https://docs.astropy.org/en/stable/index.html).

For more info on FITS files, here is a link to the FITS official [website](https://fits.gsfc.nasa.gov/). 

SIMBAD is a web-based query service from the University of Strausberg, it is a great resource for getting quick info on stars and other astronomical targets, here is the link to [Fomalhaut's SIMBAD page](https://simbad.u-strasbg.fr/simbad/sim-id?Ident=fomalhaut&NbIdent=1&Radius=2&Radius.unit=arcmin&submit=submit+id)


## What's next?

Next week we will use what we have learned about gathering TESS data and use it to plot a light curve of an exoplanetary system, in addition to exploring time series data and their applications. This will involve learning how exoplanets are detected and discover what other types of systems can be analyzed using light curves. Stay tuned!


## Acknowldegements

This notebook includes data collected with the TESS mission, obtained from the MAST data archive at the Space Telescope Science Institute (STScI). Funding for the TESS mission is provided by the NASA Explorer Program. STScI is operated by the Association of Universities for Research in Astronomy, Inc., under NASA contract NAS 5–26555.

Any published work that uses Astroquery should include a citation which can be found at [this link](https://github.com/astropy/astroquery/blob/main/astroquery/CITATION), or can be printed out in a code cell with: `astroquery.__citation__` as long as the astroquery package is imported. 

### Notebook Information:
Author: Emma Lieb

Last Updated: 05/19/2023